py36 workaround for colorama in pytest 3.2 breaks pytest-xdist on Windows

Running tests with py36 and pytest 3.2 on Windows:

```
______________ TestLooponFailing.test_looponfail_multiple_errors ______________

self = <test_looponfail.TestLooponFailing object at 0x0000024907548A20>
testdir = <Testdir local('C:\\Users\\Bruno\\AppData\\Local\\Temp\\pytest-of-Bruno\\pytest-1432\\testdir\\test_looponfail_multiple_errors0')>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x0000024907548748>

    def test_looponfail_multiple_errors(self, testdir, monkeypatch):
        modcol = testdir.getmodulecol("""
                def test_one():
                    assert 0
            """)
        remotecontrol = RemoteControl(modcol.config)
        orig_runsession = remotecontrol.runsession

        def runsession_dups():
            # twisted.trial test cases may report multiple errors.
            failures, reports, collection_failed = orig_runsession()
            print(failures)
            return failures * 2, reports, collection_failed

        monkeypatch.setattr(remotecontrol, 'runsession', runsession_dups)
>       remotecontrol.loop_once()

X:\pytest-xdist\testing\test_looponfail.py:235:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
x:\pytest-xdist\xdist\looponfail.py:113: in loop_once
    result = self.runsession()
X:\pytest-xdist\testing\test_looponfail.py:230: in runsession_dups
    failures, reports, collection_failed = orig_runsession()
x:\pytest-xdist\xdist\looponfail.py:102: in runsession
    return self.channel.receive()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <Channel id=1 closed>, timeout = None

    def receive(self, timeout=None):
        """receive a data item that was sent from the other side.
            timeout: None [default] blocked waiting.  A positive number
            indicates the number of seconds after which a channel.TimeoutError
            exception will be raised if no item was received.
            Note that exceptions from the remotely executing code will be
            reraised as channel.RemoteError exceptions containing
            a textual representation of the remote traceback.
            """
        itemqueue = self._items
        if itemqueue is None:
            raise IOError("cannot receive(), channel has receiver callback")
        try:
            x = itemqueue.get(timeout=timeout)
        except self.gateway.execmodel.queue.Empty:
            raise self.TimeoutError("no item after %r seconds" % (timeout))
        if x is ENDMARKER:
            itemqueue.put(x)  # for other receivers
>           raise self._getremoteerror() or EOFError()
E           execnet.gateway_base.RemoteError: Traceback (most recent call last):
E             File "x:\pytest-xdist\.env36\lib\site-packages\execnet\gateway_base.py", line 1076, in executetask
E               function(channel, **kwargs)
E             File "<remote exec>", line 18, in init_slave_session
E             File "x:\pytest-xdist\.env36\lib\site-packages\_pytest\config.py", line 965, in fromdictargs
E               config.parse(args, addopts=False)
E             File "x:\pytest-xdist\.env36\lib\site-packages\_pytest\config.py", line 1106, in parse
E               self._preparse(args, addopts=addopts)
E             File "x:\pytest-xdist\.env36\lib\site-packages\_pytest\config.py", line 1077, in _preparse
E               args=args, parser=self._parser)
E             File "x:\pytest-xdist\.env36\lib\site-packages\_pytest\vendored_packages\pluggy.py", line 745, in __call__
E               return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
E             File "x:\pytest-xdist\.env36\lib\site-packages\_pytest\vendored_packages\pluggy.py", line 339, in _hookexec
E               return self._inner_hookexec(hook, methods, kwargs)
E             File "x:\pytest-xdist\.env36\lib\site-packages\_pytest\vendored_packages\pluggy.py", line 334, in <lambda>
E               _MultiCall(methods, kwargs, hook.spec_opts).execute()
E             File "x:\pytest-xdist\.env36\lib\site-packages\_pytest\vendored_packages\pluggy.py", line 613, in execute
E               return _wrapped_call(hook_impl.function(*args), self.execute)
E             File "x:\pytest-xdist\.env36\lib\site-packages\_pytest\vendored_packages\pluggy.py", line 245, in _wrapped_call
E               next(wrap_controller)   # first yield
E             File "x:\pytest-xdist\.env36\lib\site-packages\_pytest\capture.py", line 39, in pytest_load_initial_conftests
E               _py36_windowsconsoleio_workaround()
E             File "x:\pytest-xdist\.env36\lib\site-packages\_pytest\capture.py", line 521, in _py36_windowsconsoleio_workaround
E               buffered = hasattr(sys.stdout.buffer, 'raw')
E           AttributeError: 'ChannelFileWrite' object has no attribute 'buffer'
```

--- a/_pytest/capture.py
+++ b/_pytest/capture.py
@@ -36,7 +36,7 @@ def pytest_addoption(parser):
 def pytest_load_initial_conftests(early_config, parser, args):
     ns = early_config.known_args_namespace
     if ns.capture == "fd":
-        _py36_windowsconsoleio_workaround()
+        _py36_windowsconsoleio_workaround(sys.stdout)
     _colorama_workaround()
     _readline_workaround()
     pluginmanager = early_config.pluginmanager
@@ -524,7 +524,7 @@ def _readline_workaround():
         pass
 
 
-def _py36_windowsconsoleio_workaround():
+def _py36_windowsconsoleio_workaround(stream):
     """
     Python 3.6 implemented unicode console handling for Windows. This works
     by reading/writing to the raw console handle using
@@ -541,13 +541,20 @@ def _py36_windowsconsoleio_workaround():
     also means a different handle by replicating the logic in
     "Py_lifecycle.c:initstdio/create_stdio".
 
+    :param stream: in practice ``sys.stdout`` or ``sys.stderr``, but given
+        here as parameter for unittesting purposes.
+
     See https://github.com/pytest-dev/py/issues/103
     """
     if not sys.platform.startswith('win32') or sys.version_info[:2] < (3, 6):
         return
 
-    buffered = hasattr(sys.stdout.buffer, 'raw')
-    raw_stdout = sys.stdout.buffer.raw if buffered else sys.stdout.buffer
+    # bail out if ``stream`` doesn't seem like a proper ``io`` stream (#2666)
+    if not hasattr(stream, 'buffer'):
+        return
+
+    buffered = hasattr(stream.buffer, 'raw')
+    raw_stdout = stream.buffer.raw if buffered else stream.buffer
 
     if not isinstance(raw_stdout, io._WindowsConsoleIO):
         return

--- a/changelog/2666.bugfix
+++ b/changelog/2666.bugfix
@@ -1,0 +1,3 @@
+Fix error on Windows and Python 3.6+ when ``sys.stdout`` has been replaced with
+a stream-like object which does not implement the full ``io`` module buffer protocol. In particular this
+affects ``pytest-xdist`` users on the aforementioned platform.

--- a/testing/test_capture.py
+++ b/testing/test_capture.py
@@ -1140,6 +1140,23 @@ def test_error_attribute_issue555(testdir):
     reprec.assertoutcome(passed=1)
 
 
+@pytest.mark.skipif(not sys.platform.startswith('win') and sys.version_info[:2] >= (3, 6),
+                    reason='only py3.6+ on windows')
+def test_py36_windowsconsoleio_workaround_non_standard_streams():
+    """
+    Ensure _py36_windowsconsoleio_workaround function works with objects that
+    do not implement the full ``io``-based stream protocol, for example execnet channels (#2666).
+    """
+    from _pytest.capture import _py36_windowsconsoleio_workaround
+
+    class DummyStream:
+        def write(self, s):
+            pass
+
+    stream = DummyStream()
+    _py36_windowsconsoleio_workaround(stream)
+
+
 def test_dontreadfrominput_has_encoding(testdir):
     testdir.makepyfile("""
         import sys

